### PR TITLE
Begin to deprecate FixInputPort overload taking a unique_ptr

### PR DESCRIPTION
**Pull request by jwnimmer-tri**
~This removes the pydrake binding for FixInputPort that takes ownership of a BasicVector, because I don't see any good reason to give pydrake more rope to hang itself with -- since the caller is not allowed to use the vec (via a retained alias) after this call, we should prevent pydrake users from accidentally doing so.~

~I noticed the lack of FixInputPort for a numpy when writing up #8856, and for #7560 I will need the by-const-reference BasicVector overload.~

_Edit: New description, after rebasing:_

This removes from `pydrake` and starts to deprecate the overload of `Context::FixInputPort` that takes ownership of a `BasicVector`.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8862)

### CODE REVIEW

**Comment by jwnimmer-tri** [comment]
+@EricCousineau-TRI for feature review, please.
\CC @sherm1 FYI on changes to `context.h`.

**Comment by EricCousineau-TRI** [comment]
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[bindings/pydrake/systems/framework_py_semantics.cc, line 77 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8862#-LDHTR-JA9ko3nG915g5:-LDHTR-K3h-_LrXX_Khb:bteclf7) ([raw file](https://github.com/robotlocomotion/drake/blob/a29e6ac4d3892757dcd97fd9155ad28d436b3484/bindings/pydrake/systems/framework_py_semantics.cc#L77)):*
> ```cpp
>            py::keep_alive<3, 1>())
>       .def("FixInputPort",
>            py::overload_cast<int, const BasicVector<T>&>(
> ```

For #7560, can I ask why this demands that we ditch the `unique_ptr` version? I am a bit tentative about completely removing this.
Why can't the downstream code just explicitly specify `my_vector.Clone()`?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8862)*

**Comment by EricCousineau-TRI** [comment]
:lgtm:, pending concerns of changing ownership semantics of `FixInputPort(..., vec: BasicVector)`

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8862#-:-LDHUIVCCtqHk46DBGr6:b-k19pmm)*

**Comment by sherm1** [comment]
Thanks, @jwnimmer-tri. :lgtm:

---

Reviewed 1 of 5 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8862#-:-LDI9WlKEcIek25CbwUi:b31makv)*

**Comment by jwnimmer-tri** [comment]
+@soonho-tri for platform review per schedule, please.

**Comment by EricCousineau-TRI** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[bindings/pydrake/systems/framework_py_semantics.cc, line 77 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8862#-LDHTR-JA9ko3nG915g5:-LDICxI662wQyV306sBc:bl22478) ([raw file](https://github.com/robotlocomotion/drake/blob/a29e6ac4d3892757dcd97fd9155ad28d436b3484/bindings/pydrake/systems/framework_py_semantics.cc#L77)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

My intent was that commit message answered this, but perhaps it was too terse.  I've amended it and added another detail to context.h API.

I've removed the unique_ptr overload from the bindings because it seemed too dangerous to use from python, and I believe that the by-const-ref method is just as good in terms of overload resolution working in pydrake?  That is, all existing callers should continue to function, unless there were mutating the `vec` after passing it in, which you aren't allowed to do anyway (it will break caching).

PTAL.
</blockquote></details>

There are enough issues in both C++ and Python as languages that attempting to address via this PR seems insufficient and kind of half-baked? Please see comment below.

---

*[systems/framework/context.h, line 255 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8862#-LDICXwW6sgLghaQxvI5:-LDICXwW6sgLghaQxvI6:b-jan2q1) ([raw file](https://github.com/robotlocomotion/drake/blob/f4100aceccdc3531bae9b2dd8cc76e08c5bdfc88/systems/framework/context.h#L255)):*
> ```objc
>   /// Identical to `FixInputPort(int, const BasicVector<T>&)` above, except
>   /// that `vec` is passed by unique_ptr instead of by-value.  The caller must
>   /// not retain any aliases to `vec` (its lifetime is not guaranteed to
> ```

I am concerned that attempting to the ownership semantics in this PR is doing too many things.

This comment is rather hard to follow; why even accept a `unique_ptr` in the first place if you're not going to retain ownership and kill the pointer at this entry point?
The comment implies that internally, the vector may be cloned, and the old value may be discarded, in which case we should completely discard the `unique_ptr` overload (and *any* similar mechanism in the code base) as this is disingenuous (and kind of a lie).

While it is true that retaining an alias and mutating will invalidate caching, if the author effectively retains a read-only alias, then there is no issue.
Additionally, because we have `get_mutable_value()`-esque mechanism to access values, and only invalidates caching at this entry point, there is *no* mechanism to fix this similar type of issue elsewhere; cutting users off at ownership transfer seems like a poor solution?

I strongly desire that this PR not try to address changing ownership semantics; I'd prefer that we discuss this more in-depth.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8862)*

**Comment by jwnimmer-tri** [comment]
-@soonho-tri -- seems like this will require a rewrite.

**Comment by jwnimmer-tri** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[bindings/pydrake/systems/framework_py_semantics.cc, line 77 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8862#-LDHTR-JA9ko3nG915g5:-LDIFTK89ZrBJBAzFE2A:b-ygzgsy) ([raw file](https://github.com/robotlocomotion/drake/blob/a29e6ac4d3892757dcd97fd9155ad28d436b3484/bindings/pydrake/systems/framework_py_semantics.cc#L77)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

There are enough issues in both C++ and Python as languages that attempting to address via this PR seems insufficient and kind of half-baked? Please see comment below.
</blockquote></details>

I guess I don't mind retaining the existing overload here (while adding the new ones), but I don't know how to unit test both of them from the python side.  What do the two different python calls look like that exercise the unique_ptr vs const-ref overloads?

---

*[systems/framework/context.h, line 255 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8862#-LDICXwW6sgLghaQxvI5:-LDIEgWt-2TRywQIVVzF:b-l3tycg) ([raw file](https://github.com/robotlocomotion/drake/blob/f4100aceccdc3531bae9b2dd8cc76e08c5bdfc88/systems/framework/context.h#L255)):*
My plan was is to kill the `unique_ptr` overload in C++ in a future PR, because as you note, this particular overload is a lie.  I didn't want to do it right away though, because it is very widely used.  I wanted to get the new overload(s) in tree first.  Maybe the problem here is the new language isn't specific enough? I could strengthen it to say "This method will become deprecated; don't use it" or alike?

In terms of the comment change here, I'd argue that I am only writing down the pre-existing implicit contract, not changing any semantics.  While I did not read every call site in tree, I sampled them and none of them appeared to be retaining this alias.  I could do a CI run that performs a Clone here, to prove it, if that helps?

I suspect your actual concern is more about the python side, so I'll reply there also.

>  ... cutting users off at ownership transfer seems like a poor solution?

In fact, I think forcing users to do `get_mutable_value`, rather than retaining long-lived mutable aliases to values within the context, is exactly what we want?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8862)*

**Comment by EricCousineau-TRI** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[bindings/pydrake/systems/framework_py_semantics.cc, line 77 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8862#-LDHTR-JA9ko3nG915g5:-LDJv8K36eVDcf5k2MLQ:bol5dst) ([raw file](https://github.com/robotlocomotion/drake/blob/a29e6ac4d3892757dcd97fd9155ad28d436b3484/bindings/pydrake/systems/framework_py_semantics.cc#L77)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I guess I don't mind retaining the existing overload here (while adding the new ones), but I don't know how to unit test both of them from the python side.  What do the two different python calls look like that exercise the unique_ptr vs const-ref overloads?
</blockquote></details>

I think my feelings on this is still pending the below discussion; however, to address the technical question, there *might* be a way to overload both `unique_ptr<T>` and `const T&` signatures, but that is stateful (i.e whether Python or C++ owns the instance), and I think less preferable to just having one of the overloads.

---

*[systems/framework/context.h, line 255 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8862#-LDICXwW6sgLghaQxvI5:-LDJtvr2Bz2Vvga0JTew:b-h24jia) ([raw file](https://github.com/robotlocomotion/drake/blob/f4100aceccdc3531bae9b2dd8cc76e08c5bdfc88/systems/framework/context.h#L255)):*
Given that this is just for `FixInputPort`, then it is true that, at the present state, changing this to implicitly clone vectors (and never transfer ownership) would work.

However, my concern is less about this particular function, but rather the broader scope, both for C++ and for Python, as I am not sure what precedent this sets for other similar APIs: we have some functions that accept `unique_ptr<T>` for ownership transfer, things that would be directly affected by caching.
Is our plan to change these things to be consistent? (i.e., to cut off any possibility of aliasing at declaration time by removing ownership transfers?)

I guess if there's a clear story for changing this as a whole, I would be a tad more OK with it, though what this tries to fix (preventing people from grabbing a pre-declaration mutable alias) is not fully preventable (one can store an alias from `get_mutable_value` immediately after - which is a bad idea of course), and is something that could be better prevented by code review (e.g. flagging storing a mutable alias in the first place)?

Before I keep on griping on something that might not actually immediately be a problem, I just want to confirm that the current `unique_ptr` overload is transferring ownership. You mentioned that:

> [...] because as you note, this particular overload is a lie [...]

Looking at the present implementation on `master`, and looking at `ContextBase::FixInputPort`, it does seem like it's doing what it says: the context is taking ownership of that particular instance; can I ask which part is a lie, in this case?

Perhaps this would be easier to go over in a f2f tomorrow or a video conference? And is there another discussion in another issue that delves further into this?

(Also, I briefly skimmed through #7560 again, but I am not sure how deprecating the `unique_ptr<T>` signature helps in this feature?)

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8862)*

**Comment by jwnimmer-tri** [comment]
I have PR'd the other changes separately.  Now, this PR is solely about (starting to) deprecate the unique_ptr overload.  Please take another look.

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/8862#-LDgLAWv4C3tiGx0aBZi:-LDgLAWv4C3tiGx0aBZj:b9o0uxs) (no related file):*
Will block on positive ack from Sherm again.

---

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/8862#-LDgLDIG9Eqn9XuDjBVL:-LDgLDIG9Eqn9XuDjBVM:b-pgr6za) (no related file):*
Will block on positive ack from Eric again.

---

*[systems/framework/context.h, line 255 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8862#-LDICXwW6sgLghaQxvI5:-LDK012s0bsXzJk82SAe:b5fq98m) ([raw file](https://github.com/robotlocomotion/drake/blob/f4100aceccdc3531bae9b2dd8cc76e08c5bdfc88/systems/framework/context.h#L255)):*
> However, my concern is less about this particular function, but rather the broader scope, both for C++ and for Python, as I am not sure what precedent this sets for other similar APIs: we have some functions that accept unique_ptr<T> for ownership transfer, things that would be directly affected by caching.

See [fixed_input_port_value.h#L75](https://github.com/RobotLocomotion/drake/blob/ee7f0def780355ae7df19740f1f6d925e6ad7726/systems/framework/fixed_input_port_value.h#L75).  The plan is for FixedInputPortValue to be sealed -- all changes go through methods that track invalidation.  This PR does not propose to change the design for the rest of the framework.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8862#-:-LDgL04MC0XcNALl7BUd:bxuckbz)*

**Comment by sherm1** [comment]
Still :lgtm:.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[systems/framework/context.h, line 250 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8862#-LDiM0ER1frhG3OIlZT4:-LDiM0ER1frhG3OIlZT5:b-gewuqh) ([raw file](https://github.com/robotlocomotion/drake/blob/9dbaf0d5e822e8360fd1952cc51aaf44cebe6023/systems/framework/context.h#L250)):*
> ```objc
>   /// FixedInputPortValue method, which will ensure that invalidation
>   /// notifications are delivered.
>   FixedInputPortValue& FixInputPort(int index, const BasicVector<T>& vec) {
> ```

BTW consider mentioning that the concrete type of the vec argument is preserved by this method.

---

*[systems/framework/context.h, line 269 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8862#-LDiMXXA22ij-CG1sKwr:-LDiMXXA22ij-CG1sKws:bc4v9jk) ([raw file](https://github.com/robotlocomotion/drake/blob/9dbaf0d5e822e8360fd1952cc51aaf44cebe6023/systems/framework/context.h#L269)):*
> ```objc
>   /// mislead users to believe that they can retain an alias of `vec` to mutate
>   /// the fixed value during a simulation.  Callers should prefer to use one of
>   /// the other overloads instead.
> ```

BTW good. We have similar misleading methods for state and parameters also. It will take some time to seal up all the leaks but this is a good place to start.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8862#-:-LDiNZAT1vJWTI0WAp9l:b-dbkjmj)*

**Comment by jwnimmer-tri** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[bindings/pydrake/systems/framework_py_semantics.cc, line 77 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8862#-LDHTR-JA9ko3nG915g5:-LDn3KKx8zoqfQA8uqp3:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a29e6ac4d3892757dcd97fd9155ad28d436b3484/bindings/pydrake/systems/framework_py_semantics.cc#L77)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

I think my feelings on this is still pending the below discussion; however, to address the technical question, there *might* be a way to overload both `unique_ptr<T>` and `const T&` signatures, but that is stateful (i.e whether Python or C++ owns the instance), and I think less preferable to just having one of the overloads.
</blockquote></details>

Done.

---

*[systems/framework/context.h, line 269 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8862#-LDiMXXA22ij-CG1sKwr:-LDn3INLAHSENnKkmt9l:b-a7rpvi) ([raw file](https://github.com/robotlocomotion/drake/blob/9dbaf0d5e822e8360fd1952cc51aaf44cebe6023/systems/framework/context.h#L269)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

It's also possible that this is overprotective and it would be fine just to prohibit holding a reference via documentation. But failure-to-invalidate-cache bugs can be extremely subtle -- typically the outdated value looks a lot like the right value during a simulation. So if it is feasible to fix our API to prevent these accidents, I'd like to do that. 
</blockquote></details>

Done.  Filed in #8904.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8862)*

**Comment by EricCousineau-TRI** [comment]
:lgtm: x 2

---

Review status: all files reviewed at latest revision, all discussions resolved.

---

*[systems/framework/context.h, line 255 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8862#-LDICXwW6sgLghaQxvI5:-LDnF6b_6aXVo8lzLvIc:b-ghdiu2) ([raw file](https://github.com/robotlocomotion/drake/blob/f4100aceccdc3531bae9b2dd8cc76e08c5bdfc88/systems/framework/context.h#L255)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

OK Satisfied given filing of issue for consistency.

---

*[systems/framework/context.h, line 269 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8862#-LDiMXXA22ij-CG1sKwr:-LDnEyZs26K6jKKqN_Ff:b9sbx6p) ([raw file](https://github.com/robotlocomotion/drake/blob/9dbaf0d5e822e8360fd1952cc51aaf44cebe6023/systems/framework/context.h#L269)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Done.  Filed in #8904.
</blockquote></details>

OK I wouldn't say that I'm 100% gung-ho for the solution, but I'm satisfied given the full scope captured in the issue. Thanks!

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8862#-:-LDnFRqt8Kqt4dhrOmsS:b4xzgsb)*
